include/oneapi/dpl/internal/distributed_ranges_impl/detail/index.hpp

```diff
@@ -0,0 +1,159 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <any>
+#include <concepts>
+#include <limits>
+#include <tuple>
+
+namespace oneapi::dpl::experimental::dr
+{
+
+namespace
+{
+template <typename T, std::size_t I, typename U = std::any>
+concept TupleElementGettable = requires(T tuple)
+{
+    {
+        std::get<I>(tuple)
+        } -> std::convertible_to<U>;
+};
+} // namespace
+
+template <typename T, typename... Args>
+concept TupleLike = requires
+{
+    typename std::tuple_size<std::remove_cvref_t<T>>::type;
+    requires std::same_as<std::remove_cvref_t<decltype(std::tuple_size_v<std::remove_cvref_t<T>>)>, std::size_t>;
+}
+&&sizeof...(Args) == std::tuple_size_v<std::remove_cvref_t<T>>&&[]<std::size_t... I>(std::index_sequence<I...>)
+{
+    return (TupleElementGettable<T, I, Args> && ...);
+}
+(std::make_index_sequence<std::tuple_size_v<std::remove_cvref_t<T>>>());
```

Comment:
I'm not sure it is important, but do we want this to work with onedpl's internal tuple type (oneapi::dpl::__internal::tuple)?  We do have an implicit conversion to `std::tuple` with the same template arguments, but I'm not sure if that works for these purposes (at compile time).

Also, this formatting looks odd.  I think I understand the requirements, but I think clang-format either needs to be run or overridden if this is it's suggestion.

Reply:
Yes, we should definitely make sure it interoperates with the `oneapi::dpl::__internal::tuple`.  However, I don't think we need to include `index.hpp` now, as it's only really needed for the matrix code, which is not included in this PR.

I just submitted a PR to remove this for now #1716.

Reply:
File was removed. Resolving.